tmt/result.py

```diff
@@ -356,119 +356,95 @@ def from_test_invocation(
-                if self.note:
-                    self.note += ', check failed'
-                else:
-                    self.note = 'check failed'
-            elif self.result == ResultOutcome.FAIL:
-                self.result = ResultOutcome.PASS
-                if self.note:
-                    self.note += ', expected failure'
-                else:
-                    self.note = 'expected failure'
-            elif self.result == ResultOutcome.PASS:
-                self.result = ResultOutcome.FAIL
-                if self.note:
-                    self.note += ', unexpected pass'
-                else:
-                    self.note = 'unexpected pass'
-            return self
-
-        # Handle individual check results
+        # First, handle individual check results
         for check in self.check:
-            check_result = CheckResultInterpret(check.result.value)
+            check_result = interpret_checks.get(check.name, CheckResultInterpret.RESPECT)
 
             if check_result == CheckResultInterpret.RESPECT:
                 if check.result == ResultOutcome.FAIL:
                     self.result = ResultOutcome.FAIL
-                    if self.note:
-                        self.note += ', check failed'
-                    else:
-                        self.note = 'check failed'
+                    self._set_note('check failed')
 
             elif check_result == CheckResultInterpret.XFAIL:
                 if check.result == ResultOutcome.PASS:
                     self.result = ResultOutcome.FAIL
-                    if self.note:
-                        self.note += ', unexpected pass in check'
-                    else:
-                        self.note = 'unexpected pass in check'
+                    self._set_note('unexpected pass in check')
                 elif check.result == ResultOutcome.FAIL:
-                    # Check failed as expected, don't change the overall result
-                    if self.note:
-                        self.note += ', expected failure in check'
-                    else:
-                        self.note = 'expected failure in check'
+                    self._set_note('expected failure in check')
 
             elif check_result == CheckResultInterpret.IGNORE:
                 continue
 
             elif check_result in [CheckResultInterpret.PASS, CheckResultInterpret.FAIL,
                                   CheckResultInterpret.INFO, CheckResultInterpret.WARN,
                                   CheckResultInterpret.ERROR, CheckResultInterpret.SKIP]:
-                self.result = ResultOutcome(check_result.value)
-                if self.note:
-                    self.note += f', check result: {check.result.value}'
-                else:
-                    self.note = f'check result: {check.result.value}'
+                # We don't promote check result to main result, but failed checks turn the
+                # main result to FAIL
```

Comment:
We don't know that yet: if the test has `result: info`, it does not matter what is the outcome of checks, they should never override the main test result, because the user clearly said "skip" is the desired and final value. We still need to resolve them, to record their final outcomes though, therefore we can't skip check results when `result: skip` is set.

I have a feeling you are getting too entangled in the problem. My advice would be to ignore the other part of the problem: focus on the interpretation of one single check result, and completely ignore what it might do to the test result. Go over all check results, and interpret every single one of them as needed. Ignore whatever their effect might be on the main result. Only after that, take `self.result` and `interpret_test`, collect the final check results, and resolve their interaction. If `test.result` is `SKIP`, `PASS`, etc., set `self.result` and `exit`. If it's `RESPECT`, compute the effect of check results, switch `self.result` to `FAIL` or `WARN`, exit. And so on. We are not in the business of saving CPU cycles, make it correct first, and then we can make it fast by introducing shortcuts. We can afford crude and readable code :)

Reply:
Maybe even one more: how about ignoring this method for now, and instead constructing a simplet method that would focus on interpretation. Feed it with ResultOutcome, ResultInterpret, map of check name/check result outome and a map of check name/check interpretation, and let it decide the final result, final check reults, and add a parametrized unit test for it for various combinations. Do not feed it with Result or CheckResuly or any structures, just enums and strings. Get that right, get it working, then we can discuss just how this process works instead of worrying sbout other structures. Once we arr happy, translate it to this method.

Reply:
Thanks again. I'll get to it tomorrow.  
> Once we arr  

🏴‍☠️

Reply:
>  I'll get to it tomorrow.

hmm, looks like a busy weekend. Monday it is.